src/number/n_unknown.jl

```diff
@@ -486,29 +507,50 @@ end
 
 end # for (rw, rew) in
 
+################################################################################
+#
+#  Coefficient ring wrapper
+#
+################################################################################
 
-function CoefficientRing(R::Nemo.Ring)
-   T = elem_type(R)
-
-   if VERSION >= v"1.8"
-      ok = ismutabletype(T)
-   else
-      ok = ismutable(R())
-   end
+CoefficientRingID = Dict{Nemo.Ring, Any}()
 
-   if R isa Nemo.Field
-      if ok
-         return N_Field{T}(R)
-      else
-         RR = FieldWrapper{typeof(R), elem_type(R)}(R)
-         return N_Field{elem_type(RR)}(RR)
-      end
-   else
-      if ok
-         return N_Ring{T}(R)
-      else
-         RR = RingWrapper{typeof(R), elem_type(R)}(R)
-         return N_Ring{elem_type(RR)}(RR)
-      end
-   end
+# We wrap everything into an N_Field or N_Ring.
+# But if the ring or the element type is mutable, we have to wrap them first
+# into something mutable (RingWrapper and FieldWrapper)
+#
+# To keep it type stable, we use a @generated function to treat being mutable
+# as a property available at compile time.
```

Comment:
Nitpick: I think mutability of types already is available at run time:
```
julia> @code_warntype Singular.ismutabletype(Int)
Variables
  #self#::Core.Const(Singular.ismutabletype)
  #unused#::Core.Const(Int64)

Body::Bool
1 ─ %1 = Base.getproperty($(Expr(:static_parameter, 1)), :mutable)::Core.Const(false)
└──      return %1
```
Based on my experience last night, I think the crucial issue is that `wrappertype` in `CoefficientRing` needs to be known at compile type, otherwise it and hence `elem_type(wrappertype)` and hence `rettype` will be not type stable...

Or maybe I got that wrong, I am not sure, my head hurts from all this (but it's still great fun to learn, thanks for looking into this, it was educational *and* entertaining to follow this PR :-)

Reply:
I think the above might also be constant propagation. I did not want to rely on this.

Reply:
I also found it entertaining and understand a bit more what `@generated` is for (and what not).